APE22.rst

```diff
@@ -0,0 +1,439 @@
+between public and internal interfaces is important for everyone, users and
+developers alike, to know what interfaces are stable and supported, and which
+are not. Consequently, Python has developed conventions for communicating what
+is public versus internal. Nominally there are three means (discussed more
+formally later):
+
+1. **Underscore Prefix**.  By convention, names that begin with an underscore
+   are considered internal.
+2. ``__all__``.  The dunder attribute ``__all__`` of a module is a list of
+   names that are considered public.
+3. **Documentation**.  Public interfaces are documented and internal interfaces
+   are not, or are specifically noted.
+
+These 3 means can be ambiguous and contradictory. Let's explore this point with
+an example. Consider an example library with the following layout::
+
+   src/
+      __init__/
+         __all__ = ["foo", "_bar"]
+
+      foo/
+         __all__ = ["_func"]
+         def func():
+            pass
+
+         def _func():
+            pass
+      
+      _bar/
+         __all__ = ['bunc']
+         def bunc():
+            pass
+
+         def _bunc():
+            pass
+
+      baz/
+         __all__ = ['qux']
+
+         def qux():
+            pass
+
+- Is ``foo`` public? Probably, but what if it's not documented?
+- Is ``func`` public? It is not in ``foo.__all__``. 
+- Is ``_func`` public?  It is prefixed with an underscore, but in ``foo.__all__``. 
+- Is ``_bar`` public? It is prefixed with an underscore, but in ``src.__all__``.
+- Is ``bunc`` public? It is in ``_bar.__all__``, but what if ``_bar`` is private?
+- Is ``_bunc`` public? Probably not, but what if it's documented?
+- Is ``baz`` public? It does not start with an underscore, but is not in ``src.__all__``.
+- Is ``qux`` public? It does not start with an underscore and is in
+  ``baz.__all__``, but what if ``baz`` is private?
+
+
+Astropy partially implements all three of the conventions: some private classes
+and functions are prefixed with an underscore, while some are not; some modules
+have an ``__all__`` attribute, and some do not;  some (debatably) internal
+interfaces are in the public documentation, like ``astropy.utils``, and some are
+not. However, Astropy does not have a uniform approach to differentiating and
+communicating public vs internal symbols, which leads to confusion for users and
+developers alike. This longstanding issue has steadily become more pressing with
+the development of assistive code tools, like auto-fill in IDEs or more recently
+with generative coding tools, like GitHub's CoPilot, which suggest code to users
+based on a corpus of code examples.  When users unintentionally use internal
+code, code tools will then suggest that internal code to other users. If we do
+not have a uniform and systematic approach to communicating what is public vs
+internal, then we cannot expect users and especially their tools to know what is
+public vs internal.
+
```

Comment:
From https://github.com/astropy/astropy/issues/15169, I recall another source of issues that has bitten astropy in the past: I/O registration. When deep sub-modules look public, e.g. `astropy.cosmology.flrw.lambdacdm.LambdaCDM` then that is generally used as the class pathway in I/O rather than the better `astropy.cosmology.LambdaCDM`. When the class is moved it breaks the I/O and needs a whole backwards-compatibility logic to support files using the previous path. Cleaning up public vs internal, e.g. to `astropy.cosmology._flrw.lambdacdm.LambdaCDM` would make this an obviously bad way to serialize the class and code authors would get it "right" (`astropy.cosmology.LambdaCDM`) the first time. Clear public API makes for more stable code.

probably deserves a few sentences.

Reply:
This is slightly orthogonal to the issue here: really, one needs to set `__module__` to avoid this - registration just followed that, not a choice made by coders. Note that `numpy` in fact does this - which partially is annoying, since one no longer knows where the function is defined.

Reply:
For me the pain point was YAML serialization because I used ``f"{class.__module__}.{class.__qualname__}"``, which looks reasonable as code, but led to the problems described above. I would not have pushed the code as I originally did if, when making the tests, I had noticed a private path in the way.

Reply:
Yes, that one is an issue for examples and `sphinx` too - I really dislike that when we made `representations` a module (which was definitely a good idea!), we had to make so many updates to the docs. It would also be nice to use `regex` for things like [`__construct_mixin_classes`](https://github.com/astropy/astropy/blob/291414a3b600a12738762b7c50bfc6709f3cbf13/astropy/table/serialize.py#L25)

For me, this is still separate from the APE, since this has annoyed me amply already! But I can see that coming from an environment where files with leading underscores have more meaning, you could have avoided the issues.

Reply:
Agreed we can leave it out of this APE. I think the point I made "If, when making the tests, I had noticed a private path I would have fixed the problem from the start." means adopting this APE will automatically force most of these I/O issues to be fixed